celery/concurrency/workhorse.py

```diff
@@ -0,0 +1,265 @@
+                        pass
+                _set_current_process(self)
+
+                # Re-init logging system.
+                # Workaround for http://bugs.python.org/issue6721/#msg140215
+                # Python logging module uses RLock() objects which are broken
+                # after fork. This can result in a deadlock (Celery Issue #496).
+                loggerDict = logging.Logger.manager.loggerDict
+                logger_names = list(loggerDict.keys())
+                logger_names.append(None)  # for root logger
+                for name in logger_names:
+                    if not name or not isinstance(loggerDict[name],
+                                                  logging.PlaceHolder):
+                        for handler in logging.getLogger(name).handlers:
+                            handler.createLock()
+                logging._lock = threading.RLock()
+
+                util._finalizer_registry.clear()
+                util._run_after_forkers()
+                try:
+                    target(*args, **kwargs)
+                    exitcode = 0
+                finally:
+                    util._exit_function()
+            except SystemExit as exc:
+                if not exc.args:
+                    exitcode = 1
+                elif isinstance(exc.args[0], int):
+                    exitcode = exc.args[0]
+                else:
+                    sys.stderr.write(str(exc.args[0]) + '\n')
+                    exitcode = 0 if isinstance(exc.args[0], str) else 1
+            except:
+                exitcode = 1
+                if not util.error('Process %s', self.name, exc_info=True):
+                    import traceback
+                    sys.stderr.write('Process %s:\n' % self.name)
+                    traceback.print_exc()
+            finally:
+                util.info('Process %s exiting with exitcode %d', self.pid, exitcode)
+                _maybe_flush(sys.stdout)
+                _maybe_flush(sys.stderr)
+                _maybe_flush(sys.__stdout__)
+                _maybe_flush(sys.__stderr__)
+                os._exit(exitcode)
+
+class TaskPool(BasePool):
+    sigfd = None
+    sigfh = None
+    workers = ()
+    uses_semaphore = True
+    sem = None
+
+    def on_start(self):
+        self.sem = self.options['semaphore']
+        self.workers = {}
+        self.sigfd = signalfd.signalfd(0, [signal.SIGCHLD], signalfd.SFD_NONBLOCK | signalfd.SFD_CLOEXEC)
+        self.sigfh = os.fdopen(self.sigfd, 'rb')
+        signalfd.sigprocmask(signalfd.SIG_BLOCK, [signal.SIGCHLD])
+
+    def register_with_event_loop(self, hub):
+        hub.add_reader(self.sigfd, self.on_sigchld)
+
+    def on_close(self):
+        self.sigfh.close()
+        signalfd.sigprocmask(signalfd.SIG_BLOCK, [])
+
+    def on_sigchld(self):
+        pending = {}
+
+        si = signalfd_siginfo()
+        while True:
+            try:
+                if not self.sigfh.readinto(si):
```

Comment:
I understand why you are doing this but I'm really not 100% sure that is will be performant when using PyPy.